woocommerce/woocommerce
# Core block-based template PHP API (adding blocks to templates)

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds an internal `BlockTemplate` class, along with supporting classes.

Non-internal (available to extension developers) interfaces for the following are exposed:

* `BlockTemplateInterface`
* `BlockContainerInterface`
* `BlockInterface`

This PR includes a number of unit tests to both verify the implementation and show how the API can be used (**_roughly 60% of the lines of code in this PR is unit tests_**).

Follow-up PRs will add support for a hook to allow the block-based product form template to be modified by extensions, as well as add additional capabilities to the API.

Closes #39418.



### How to test the changes in this Pull Request:




1. Run the unit tests and verify they pass:

```
pnpm --filter=woocommerce env:test
pnpm test:unit:env --filter 'Automattic\\WooCommerce\\Tests\\Internal\\Admin\\BlockTemplates'
```

(Or, just verify that the tests passed on this PR).



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
